File: sdk/python/src/hr_agent_datagraph/validation.py
"""JSON Schema validation for HR data files."""

from __future__ import annotations

from pathlib import Path
from typing import Any

try:
    import jsonschema
    HAS_JSONSCHEMA = True
except ImportError:
    HAS_JSONSCHEMA = False


class ValidationResult:
    """Result of a schema validation check."""

    def __init__(self, errors: list[str] | None = None):
        self.errors = errors or []

    @property
    def is_valid(self) -> bool:
        return len(self.errors) == 0

    def __repr__(self) -> str:
        if self.is_valid:
            return "ValidationResult(valid=True)"
        return f"ValidationResult(valid=False, errors={self.errors})"


def validate(data: dict[str, Any], schema_name: str) -> ValidationResult:
    """Validate data against a named schema.

    Args:
        data: The data to validate
        schema_name: Schema file name (without path), e.g. "offboarding-taxonomy"

    Returns:
        ValidationResult with any errors found.
    """
    if not HAS_JSONSCHEMA:
        return ValidationResult(["jsonschema package not installed. Install with: pip install jsonschema"])

    schema_root = Path(__file__).resolve().parent.parent.parent.parent.parent / "schemas"
    schema_path = schema_root / f"{schema_name}.schema.json"

    if not schema_path.exists():
        return ValidationResult([f"Schema not found: {schema_path}"])

    import json
    with open(schema_path, encoding="utf-8") as f:
        schema = json.load(f)

    errors = []
    try:
        jsonschema.validate(instance=data, schema=schema)
    except jsonschema.ValidationError as e:
        errors.append(f"{'.'.join(str(p) for p in e.absolute_path) or '<root>'}: {e.message}")
    except jsonschema.SchemaError as e:
        errors.append(f"Schema error: {e.message}")

    return ValidationResult(errors)

File: sdk/typescript/src/validation.ts
/**
 * JSON Schema validation for HR data files.
 */

import { readFileSync, existsSync } from "node:fs";
import { join, resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

/**
 * Validate data against a named schema.
 * Note: For full validation, install `ajv` as a peer dependency.
 *
 * @param data - The data to validate
 * @param schemaName - Schema name (without path), e.g. "offboarding-taxonomy"
 * @returns ValidationResult with any errors found
 */
export function validate(data: unknown, schemaName: string): ValidationResult {
  const schemaRoot = resolve(__dirname, "../../../schemas");
  const schemaPath = join(schemaRoot, `${schemaName}.schema.json`);

  if (!existsSync(schemaPath)) {
    return { valid: false, errors: [`Schema not found: ${schemaPath}`] };
  }

  // Basic structural validation (full validation requires ajv)
  try {
    const schema = JSON.parse(readFileSync(schemaPath, "utf-8"));
    const errors: string[] = [];

    if (schema.required && typeof data === "object" && data !== null) {
      for (const field of schema.required) {
        if (!(field in (data as Record<string, unknown>))) {
          errors.push(`Missing required field: ${field}`);
        }
      }
    }

    return { valid: errors.length === 0, errors };
  } catch (err) {
    return {
      valid: false,
      errors: [`Validation error: ${err instanceof Error ? err.message : String(err)}`],
    };
  }
}
